src/unif.ml

```diff
@@ -237,7 +237,7 @@ and solve_aux : term -> term -> problems -> unif_constrs = fun t1 t2 p ->
    injective. Currently, it only handles a specific case: when s is the
    builtin P. *)
   let solve_inj s ts v =
-    if !(s.sym_rules) = [] then error ()
+    if !(s.sym_rules) = [] then add_to_unsolved ()
```

Comment:
This modification is not directly related to SR. If you think that this is a bug, this should done in a separate oneline PR.

Reply:
In the algorithm for checking SR, we introduce new function symbols (c_x, t_x) which are not like user-defined symbols. When we attempt to unify (c_x, t), where t is not headed by a symbol, we shouldn't raise an error like we did here. Maybe we should check if s is one of these special symbols but the code will become more complicated by doing so.

Reply:
Why shouldn't we raise an error? How can you have c_x = t if c_x is undefined and t is not headed by a function symbol?

Reply:
To check SR for the rule l -> r, we check Γ |- lθ : T => Γ |- rθ : T. Here c_x actually corresponds to the term xθ and t_x to the type of any occurrence of xθ in the derivation of Γ |- lθ : T as I understand from the algorithm.

Reply:
Yes, but Γ |- rθ : T has to be done in a system where l->r has not been added yet. It would be strange to need l->r (in type conversions) to check that l->r preserves typing...

Reply:
Yes, this has been corrected but I think you might have pasted this comment in the wrong place.

Reply:
What do you mean? Your last commit is on May 3...

Reply:
I think your last comment refers to the modification I made in handle.ml, but not this part of code...

Reply:
Yes, there seems to be some confusion. So, what has been corrected?

Reply:
There are two points to clarify:
1. in handle: add a rule in the signature AFTER successful SR check only
2. in unif: solve_inj should fail if the head symbol has no rule

Reply:
https://github.com/Deducteam/lambdapi/pull/200#discussion_r290650646 This has been corrected but I haven't made the commit yet.

Reply:
Ok, thanks.

Reply:
It would be good to commit more often so that I know what is fixed or not, and can close some of the discussions.

Reply:
For the point 2, see https://github.com/Deducteam/lambdapi/pull/200#discussion_r290701911 or consider the following example:
symbol eta : (nat => nat) => nat => nat
rule eta &g -> \x. &g[x]
In the algorithm for checking SR, we first replace the metavariable &g with a new function symbol c_g of type t_g  and then try to infer the type of modified LHS, i.e. eta c_g. The type inference of LHS will generate the constraint t_g ~ nat => nat, which corresponds to the constraint T(xθ) ~ nat => nat obtained from the derivation Γ |- lθ : T (T(xθ) is the type of any occurrence of xθ in this derivation).

Reply:
This rule should be rejected at the scope level because g is of arity 0 in the LHS and 1 in the RHS. The arity should be constant.
Now, yes, when trying to type the LHS, you get the constraint t_g = nat=>nat. Then, it means that you should check that the RHS (say g) is of type nat=>nat using the rule t_g --> nat=>nat, but not the rule eta g --> g itself!

Reply:
I think there must be some misunderstanding. When we try to type LHS, we first try to unify t_g and nat => nat. This is the origin of the constraint t_g = nat => nat. The original version of the unification function raises a fatal error in this case. However, instead of raising an error, we should add this equation into the unsolved part. I was not reasoning about using the rule itself to check SR...

Reply:
I see. But my idea was initially to proceed differently:
Given a rule l --> r,
1. in l, replace pattern variables by metavariables, giving l'
2. apply the standard type inference algorithm on l' (because unification acts on metavariables) giving a type T and a set of constraints E
3. build a map M from the (unsolved) metavariables of l' or their types to new symbols
4. complete M(E) into a convergent rewrite system R
5. try to check that M(r) has type M(T) modulo R